### PR TITLE
Add Windows InitOnceBeginInitialize and InitOnceComplete shims

### LINKED ISSUE #2595: can't call foreign function: InitOnceBeginInitialize when running on Windows
My crate [`aligners`](https://github.com/v0ldek/aligners) got a failure in a scheduled Miri pipeline ([job](https://github.com/V0ldek/aligners/actions/runs/3246804722/jobs/5333903048)) absent any changes to the crate's code on the nightly version available in the night between 13th and 14th of October.

Three of the pipeline targets seem to be failing:
- x86_64-pc-windows-gnu
- i686-pc-windows-gnu
- i686-pc-windows-msvc

Surprisingly, x86_64-pc-windows-msvc seems to be working.

Here's the full error:

```
error: unsupported operation: can't call foreign function: InitOnceBeginInitialize
Error:   --> C:\Users\runneradmin\.rustup\toolchains\nightly-x86_[64](https://github.com/V0ldek/aligners/actions/runs/3246804722/jobs/5333903357#step:4:65)-pc-windows-msvc\lib\rustlib\src\rust\library\std\src\sys\windows\thread_local_key.rs:88:21
   |
88 |             let r = c::InitOnceBeginInitialize(self.once.get(), 0, &mut pending, ptr::null_mut());
   |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ can't call foreign function: InitOnceBeginInitialize
   |
   = help: this is likely not a bug in the program; it indicates that the program performed an operation that the interpreter does not support
   = note: BACKTRACE:
   = note: inside `std::sys::windows::thread_local_key::StaticKey::init` at C:\Users\runneradmin\.rustup\toolchains\nightly-x86_64-pc-windows-msvc\lib\rustlib\src\rust\library\std\src\sys\windows\thread_local_key.rs:88:21
   = note: inside `std::sys::windows::thread_local_key::StaticKey::key` at C:\Users\runneradmin\.rustup\toolchains\nightly-x86_64-pc-windows-msvc\lib\rustlib\src\rust\library\std\src\sys\windows\thread_local_key.rs:79:18
   = note: inside `std::sys::windows::thread_local_key::StaticKey::get` at C:\Users\runneradmin\.rustup\toolchains\nightly-x86_64-pc-windows-msvc\lib\rustlib\src\rust\library\std\src\sys\windows\thread_local_key.rs:73:24
   = note: inside `std::thread::__OsLocalKeyInner::<std::cell::RefCell<std::option::Option<std::sys_common::thread_info::ThreadInfo>>>::get::<[closure@std::sys_common::thread_info::THREAD_INFO::__getit::{closure#0}]>` at C:\Users\runneradmin\.rustup\toolchains\nightly-x86_64-pc-windows-msvc\lib\rustlib\src\rust\library\std\src\thread\local.rs:1080:32
   = note: inside `std::sys_common::thread_info::THREAD_INFO::__getit` at C:\Users\runneradmin\.rustup\toolchains\nightly-x86_64-pc-windows-msvc\lib\rustlib\src\rust\library\std\src\thread\local.rs:271:21
   = note: inside `std::thread::LocalKey::<std::cell::RefCell<std::option::Option<std::sys_common::thread_info::ThreadInfo>>>::try_with::<[closure@std::sys_common::thread_info::set::{closure#0}], ()>` at C:\Users\runneradmin\.rustup\toolchains\nightly-x86_64-pc-windows-msvc\lib\rustlib\src\rust\library\std\src\thread\local.rs:445:32
   = note: inside `std::thread::LocalKey::<std::cell::RefCell<std::option::Option<std::sys_common::thread_info::ThreadInfo>>>::with::<[closure@std::sys_common::thread_info::set::{closure#0}], ()>` at C:\Users\runneradmin\.rustup\toolchains\nightly-x86_64-pc-windows-msvc\lib\rustlib\src\rust\library\std\src\thread\local.rs:422:9
   = note: inside `std::sys_common::thread_info::set` at C:\Users\runneradmin\.rustup\toolchains\nightly-x86_64-pc-windows-msvc\lib\rustlib\src\rust\library\std\src\sys_common\thread_info.rs:42:5
   = note: inside `std::rt::init` at C:\Users\runneradmin\.rustup\toolchains\nightly-x86_64-pc-windows-msvc\lib\rustlib\src\rust\library\std\src\rt.rs:105:9
   = note: inside closure at C:\Users\runneradmin\.rustup\toolchains\nightly-x86_64-pc-windows-msvc\lib\rustlib\src\rust\library\std\src\rt.rs:147:42
   = note: inside `std::panicking::r#try::do_call::<[closure@std::rt::lang_start_internal::{closure#1}], ()>` at C:\Users\runneradmin\.rustup\toolchains\nightly-x86_64-pc-windows-msvc\lib\rustlib\src\rust\library\std\src\panicking.rs:483:40
   = note: inside `std::panicking::r#try::<(), [closure@std::rt::lang_start_internal::{closure#1}]>` at C:\Users\runneradmin\.rustup\toolchains\nightly-x86_64-pc-windows-msvc\lib\rustlib\src\rust\library\std\src\panicking.rs:447:19
   = note: inside `std::panic::catch_unwind::<[closure@std::rt::lang_start_internal::{closure#1}], ()>` at C:\Users\runneradmin\.rustup\toolchains\nightly-x86_64-pc-windows-msvc\lib\rustlib\src\rust\library\std\src\panic.rs:137:14
   = note: inside `std::rt::lang_start_internal` at C:\Users\runneradmin\.rustup\toolchains\nightly-x86_64-pc-windows-msvc\lib\rustlib\src\rust\library\std\src\rt.rs:147:5
   = note: inside `std::rt::lang_start::<()>` at C:\Users\runneradmin\.rustup\toolchains\nightly-x86_64-pc-windows-msvc\lib\rustlib\src\rust\library\std\src\rt.rs:1[65](https://github.com/V0ldek/aligners/actions/runs/3246804722/jobs/5333903357#step:4:66):17
```

There's none of my code in this trace. The tests failing [are here](https://github.com/V0ldek/aligners/blob/main/src/lib.rs#L159), but they don't seem to be even running when this fails. Rerunning the actions causes the same result, so it doesn't look like a transient error.

### PR DESCRIPTION
**Pull request by DrMeepster**
Fixes #2595

### CODE REVIEW

File: src/concurrency/sync.rs [hunk 1]
```diff
@@ -149,13 +150,68 @@ struct FutexWaiter {
...
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+/// The current status of a one time initialization.
+pub enum InitOnceStatus {
+    Uninitialized,
+    Begun,
+    Complete,
+}
+
+impl Default for InitOnceStatus {
```
Comment (Nilstrieb) [inline src/concurrency/sync.rs:177]: Looks like a great use case for the new `#[default]` derive attribute :)

File: src/shims/windows/sync.rs [hunk 2]
```diff
@@ -25,17 +29,38 @@ fn srwlock_get_or_create_id<'mir, 'tcx: 'mir>(
...
-    })
+    }
+
+    fn srwlock_get_or_create_id(
+        &mut self,
+        lock_op: &OpTy<'tcx, Provenance>,
+    ) -> InterpResult<'tcx, RwLockId> {
+        let this = self.eval_context_mut();
+        this.rwlock_get_or_create(|ecx, next_id| {
+            Ok(ecx.get_or_create_id(next_id.to_u32_scalar(), lock_op)?.map(RwLockId::from_u32))
```
Comment (oli-obk) [inline src/shims/windows/sync.rs:42]: We could auto-implement a trait for ids that has these conversion methods and then only have a single function that is generic over the trait

File: src/concurrency/sync.rs (hunk 1 above)
Comment (DrMeepster) [reply src/concurrency/sync.rs:177]: Oh, it looks like it's stable. I thought it was still unstable.

File: rust-version [hunk 3]
```diff
@@ -1 +1 @@
-538f118da1409759ba198acc0ff62070bc6d2dce
+a24a020e6d926dffe6b472fc647978f92269504e
```
Comment (RalfJung) [inline rust-version:1]: If you only bump to edabf59ca4646b3fc1a961c26431215001043f6a, does that help for CI?

File: src/concurrency/sync.rs [hunk 4]
```diff
@@ -149,18 +158,99 @@ struct FutexWaiter {
     bitset: u32,
 }
 
+declare_id!(InitOnceId);
+
+struct InitOnceWaiter<'mir, 'tcx> {
+    thread: ThreadId,
+    callback: Box<dyn MachineCallback<'mir, 'tcx> + 'tcx>,
+}
```
Comment (RalfJung) [inline src/concurrency/sync.rs:166]: Please add doc comments. In particular I am wondering what the callback is doing.

File: src/concurrency/sync.rs [hunk 5]
```diff
@@ -581,4 +707,144 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriInterpCxExt<'mir, 'tcx> {
...
+        assert_eq!(
+            init_once.status,
+            InitOnceStatus::Begun,
+            "failing already completed or uninit init once"
+        );
+
+        // Each complete happens-before the end of the wait
+        if let Some(data_race) = &this.machine.data_race {
+            data_race.validate_lock_release(&mut init_once.data_race, current_thread);
+        }
```
Comment (RalfJung) [inline src/concurrency/sync.rs:824]: This is a 'fail'. I don't think it should induce any synchronization.

File: src/concurrency/sync.rs [hunk 6]
```diff
@@ -581,4 +707,144 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriInterpCxExt<'mir, 'tcx> {
...
+
+        // the behavior of failing the initialization is left vague by the docs
+        // it had to be determined experimentally
```
Comment (RalfJung) [inline src/concurrency/sync.rs:827]: So what was the result of those experiments? Please write the text you would have liked to see in the docs.

Also, which docs? This is not a 'windows' file.

File: src/shims/unix/sync.rs [hunk 7]
```diff
@@ -435,7 +354,7 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriInterpCxExt<'mir, 'tcx> {
         let this = self.eval_context_mut();
 
         let kind = mutex_get_kind(this, mutex_op)?;
-        let id = mutex_get_or_create_id(this, mutex_op)?;
+        let id = this.mutex_get_or_create_id(mutex_op, 4)?;
```
Comment (RalfJung) [inline src/shims/unix/sync.rs:357]: Please define some `const` for this 4. It used to be in one place only but your refactor lead to duplication. (Same for the other offsets in this file and the Windows file.)

File: src/shims/windows/sync.rs [hunk 8]
```diff
@@ -135,4 +103,112 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriInterpCxExt<'mir, 'tcx> {
...
+            }
+        }
+
+        impl<'mir, 'tcx> MachineCallback<'mir, 'tcx> for Callback<'tcx> {
+            fn call(&self, this: &mut MiriInterpCx<'mir, 'tcx>) -> InterpResult<'tcx> {
+                let pending = match this.init_once_status(self.init_once_id) {
+                    InitOnceStatus::Uninitialized =>
+                        unreachable!("status should have either been set to begun or complete"),
+                    InitOnceStatus::Begun => this.eval_windows("c", "TRUE")?,
+                    InitOnceStatus::Complete => this.eval_windows("c", "FALSE")?,
```
Comment (RalfJung) [inline src/shims/windows/sync.rs:148]: Interesting, I didn't know about these constants. But we should be consistent -- I think all the other Windows shims use `Scalar::from_i32` to create their BOOL literals. If we're going with eval for this, can you make a separate PR changing the existing code?

File: src/shims/windows/sync.rs [hunk 9]
```diff
@@ -135,4 +103,112 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriInterpCxExt<'mir, 'tcx> {
...
+            InitOnceStatus::Uninitialized => {
+                this.init_once_begin(id);
+                this.write_scalar(this.eval_windows("c", "TRUE")?, &pending_place)?;
+            }
+            InitOnceStatus::Begun =>
+                this.init_once_enqueue_and_block(
+                    id,
+                    active_thread,
+                    Box::new(Callback { init_once_id: id, pending_place }),
+                ),
```
Comment (RalfJung) [inline src/shims/windows/sync.rs:167]: Please move the callback into this match arm. It is confusing to read all that callback code and then realize it is only sometimes relevant.

**Review by RalfJung — COMMENTED** [review]
Thanks, this mostly looks great!

File: rust-version (hunk 3 above)
Comment (DrMeepster) [reply rust-version]: That worked, thanks.

File: src/concurrency/sync.rs (hunk 6 above)
Comment (DrMeepster) [reply src/concurrency/sync.rs:827]: It actually doesn't matter because it is clearly documented, and I just missed it. Whoops.

**Comment by RalfJung** [comment]
In the interest of having our CI fixed, I plan to take over this PR later today so that it can land ASAP. I hope that's okay.
